tomviz/ExportDataReaction.cxx

```diff
@@ -0,0 +1,280 @@
+  ToType* a = static_cast<ToType*>(outArray->GetVoidPointer(0));
+  for (int i = 0; i < nComps * nTuples; ++i) {
+    a[i] = static_cast<ToType>(d[i]);
+  }
+}
+
+template <typename FromType>
+void convertToUnsignedChar(vtkDataArray* outArray, int nComps, int nTuples,
+                           void* data)
+{
+  convert<FromType, unsigned char>(outArray, nComps, nTuples, data);
+}
+}
+
+bool ExportDataReaction::exportData(const QString& filename)
+{
+  auto server = pqActiveObjects::instance().activeServer();
+
+  vtkSmartPointer<vtkDataObject> data = this->m_module->getDataToExport();
+
+  if (!server) {
+    qCritical("No active server located.");
+    return false;
+  }
+
+  QFileInfo info(filename);
+  if (info.suffix() == "emd") {
+    EmdFormat writer;
+    auto image = vtkImageData::SafeDownCast(data);
+    if (!image || !writer.write(filename.toLatin1().data(), image)) {
+      qCritical() << "Failed to write out data.";
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  auto writerFactory = vtkSMProxyManager::GetProxyManager()->GetWriterFactory();
+
+  vtkSMSessionProxyManager* pxm =
+    vtkSMProxyManager::GetProxyManager()->GetActiveSessionProxyManager();
+
+  vtkSmartPointer<vtkSMSourceProxy> producer;
+  producer.TakeReference(vtkSMSourceProxy::SafeDownCast(
+    pxm->NewProxy("sources", "TrivialProducer")));
+  vtkSmartPointer<vtkTrivialProducer> trivialProducer =
+    vtkTrivialProducer::SafeDownCast(producer->GetClientSideObject());
+  trivialProducer->SetOutput(data);
+  trivialProducer->UpdateInformation();
+  trivialProducer->Update();
+  producer->UpdatePipeline();
+
+  vtkSmartPointer<vtkSMProxy> proxy;
+  proxy.TakeReference(
+    writerFactory->CreateWriter(filename.toLatin1().data(), producer));
+  auto writer = vtkSMSourceProxy::SafeDownCast(proxy);
+  if (!writer) {
+    qCritical() << "Failed to create writer for: " << filename;
+    return false;
+  }
+
+  // Convert to a data format the file type supports
+  const char* writerName = writer->GetClientSideObject()->GetClassName();
+  auto imageData =
+    vtkImageData::SafeDownCast(trivialProducer->GetOutputDataObject(0));
+  if (imageData) {
+    auto imageType = imageData->GetPointData()->GetScalars()->GetDataType();
+    if (strcmp(writerName, "vtkTIFFWriter") == 0 && imageType == VTK_DOUBLE) {
+      vtkNew<vtkImageData> fImage;
+      fImage->DeepCopy(imageData);
+      ConvertToFloatOperator convertFloat;
+      convertFloat.applyTransform(fImage);
+
+      trivialProducer->SetOutput(fImage.Get());
+      trivialProducer->UpdateInformation();
+      trivialProducer->Update();
+      producer->UpdatePipeline();
+    }
+
+    if ((strcmp(writerName, "vtkPNGWriter") == 0 &&
+         (imageType != VTK_UNSIGNED_CHAR || imageType != VTK_UNSIGNED_SHORT)) ||
+        (strcmp(writerName, "vtkJPEGWriter") == 0 &&
+         imageType != VTK_UNSIGNED_CHAR)) {
+      std::cout << "File type does not support the current data type, converting "
```

Comment:
We can leave this for now, but I think it is much better to pop up a message box, with an option to silence forever that write to the standard out. If you want to write to terminal qDebug/qWarning would be more consistent here.